SCAutolib/models/gui.py

```diff
@@ -0,0 +1,304 @@
+        # keyboard.write types nothing if the delay is not set
+        self.write = functools.partial(kb_decorator(keyboard.write), delay=0.1)
+        self.send = kb_decorator(keyboard.send)
+        # For some reason the first keypress is never sent
+        # So this effectively does nothing
+        keyboard.send('enter')
+
+
+class GUI():
+    def __init__(self, wait_time=5):
+        self.WAIT_TIME = wait_time
+        self.GDM_INIT_TIME = 10
+        # Create the directory for screenshots
+        self.screenshot_directory = '/tmp/SC-tests/' + str(int(time()))
+        os.makedirs(self.screenshot_directory, exist_ok=True)
+
+        self.mouse = Mouse()
+        self.kb = KB(self.WAIT_TIME)
+
+    def __enter__(self):
+        self.screen = Screen(self.screenshot_directory)
+        # By restarting gdm, the system gets into defined state
+        run(['systemctl', 'restart', 'gdm'], check=True)
+        # Cannot screenshot before gdm starts displaying
+        # This would break the display
+        sleep(self.GDM_INIT_TIME)
+
+        return self
+
+    def __exit__(self, type, value, traceback):
+        # Gather the screenshots, generate report
+        pass
+
+    @staticmethod
+    def _image_to_data(filename):
+        """
+        Convert screenshot into dataframe of words with their coordinates.
+        """
+        UPSCALING_FACTOR = 2
+
+        image = cv2.imread(filename)
+        grayscale = cv2.cvtColor(image, cv2.COLOR_BGR2GRAY)
+        upscaled = cv2.resize(grayscale,
+                              dsize=None,
+                              fx=UPSCALING_FACTOR,
+                              fy=UPSCALING_FACTOR,
+                              interpolation=cv2.INTER_LANCZOS4)
+        _, binary = cv2.threshold(upscaled, 120, 255, cv2.THRESH_BINARY_INV)
+        image_data_str = pytesseract.image_to_data(binary)
+        df = pd.read_csv(StringIO(image_data_str),
+                         sep='\t', lineterminator='\n')
+
+        yres, xres = binary.shape[:2]
+        df[['left', 'width']] /= xres
+        df[['top', 'height']] /= yres
+
+        logger.debug(df)
+        return df
+
+    @staticmethod
+    def _images_same(filename1: str, filename2: str):
+        """Compare two images, return True if they are completely identical."""
+        im1 = cv2.imread(filename1)
+        im2 = cv2.imread(filename2)
+
+        # Is the resolution the same
+        if im1.shape != im2.shape:
+            return False
+
+        # Check if value of every pixel is the same
+        if np.bitwise_xor(im1, im2).any():
+            return False
+
+        return True
+
+    def click_on(self, key: str, timeout: float = 30):
+        """Clicks matching word on the screen.
+
+        """
+        logger.info(f"Trying to find key='{key}' to click on.")
+
+        end_time = time() + timeout
+        item = None
+        first_scr = None
+
+        # Repeat screenshotting, until the key is found
+        while time() < end_time:
```

Comment:
If there is no match, this method will produce many screenshots with the same content. I would consider comparing images with _images_same method, removing duplicit images and resetting counter. I would consider implementing such mechanism in the Screen class.